File: 2015/day_01/cpp/attempt_2.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ifstream in("../input.txt");
    string s; in >> s;
    int f=0,p=0;
    for(int i=0;i<s.size();i++){
        f += s[i]=='('?1:-1;
        if(!p && f==-1) p=i+1;
    }
    cout<<f<<" "<<p<<"\n";
}

File: 2015/day_01/haskell/attempt_2.hs
main :: IO ()
main = do
    s <- readFile "../input.txt"
    let floors = scanl step 0 s
        final  = last floors
        pos    = length (takeWhile (/= -1) floors)
    putStrLn $ "Part 1: Final floor = " ++ show final
    putStrLn $ "Part 2: First basement position = " ++ show pos

step :: Int -> Char -> Int
step lvl '(' = lvl + 1
step lvl ')' = lvl - 1
step lvl _   = lvl
